### PR TITLE
[Consolidated Channel] Update target pod IPs when dispatcher pods scale before the subscription is marked ready 

### PR DESCRIPTION
**Pull request by devguyio**
## Background
The existing dispatcher pods Prober acts like this:
1. `IsReady(sub)` -> get list of target pods, cache the state (including the list of pods), and start probing those pods.
3. Probing collects ready partitions and stores them in the cached target state
4. After each dispatcher pod partitions are cached, check if we collected all partitions we are looking for and if so, mark the cached targetState as ready.  

There's a corner case the manifests itself as follows:
1. `IsReady(sub)` -> get list of target pods, cache the state (including the list of pods), and start probing those pods.
2.  Dispatcher pods scale up before the cached targetState becomes ready. 
3.  No new probes are created for the new pods.
4. The existing probing returns only a subset of the partitions, because not all dispatcher pods are probed for this targetState.

This is a fix for the behavior [described here](https://github.com/knative-sandbox/eventing-kafka/pull/476#issuecomment-809156584)

### Kafka Controller 
1. Refresh prober pods when dispatcher pods are added or deleted.

### Prober
1. Refactor the code to be more readible by breaking down `IsReady` and getting rid of non-readable anonymous function and logic.
2. Consider an in-flight targetState outdated if the list of initial probed pods change from the current, forget the state and reprobe new pods.
3. Add `RefreshPodProbing()` that will eject outdated in-flight target states, and reprobe the new pods instead.

### CODE REVIEW

File: pkg/channel/consolidated/reconciler/controller/kafkachannel.go
Comment (lberk) [inline pkg/channel/consolidated/reconciler/controller/kafkachannel.go:244]: why wouldn't this be considered a new error type and then we can warn on that?

Comment (devguyio) [reply pkg/channel/consolidated/reconciler/controller/kafkachannel.go:244]: can you explain more? as in consider the case where not all subscribers were reconciled an error and check the error returned by `reconcileSubscribers` instead of a bool and error?

Comment (lberk) [reply pkg/channel/consolidated/reconciler/controller/kafkachannel.go:244]: right.  We're mixing the return value across multiple variables at this point.  For example, the fact we sometimes return `reconciled` and othertimes elect to return a constant (`false`) indicates to me, that something is amiss with the logic as it currently stands.

let's keep the single return value (of type `error`) and have the control flow dependant on the actual error.

File: pkg/channel/consolidated/reconciler/controller/kafkachannel.go
Comment (lberk) [inline pkg/channel/consolidated/reconciler/controller/kafkachannel.go:266]: realistically, if we know that we haven't properly reconciled here, why are we even continuing down the rest of the function?

**Comment by lberk** [comment]
https://github.com/knative-sandbox/eventing-kafka/blob/3784fe05d38212764c6130f3db0b7d61b8bb4c29/pkg/channel/consolidated/status/status.go#L186-L221
tbh I'm finding this lambda block really hard to review now.  I'd appreciate it if you could abstract this into its own function with named return variables so I can understand the logic a bit better.  Especially because now we're dealing with prematurely unlocking and re-locking.

File: pkg/channel/consolidated/status/status.go
Comment (lberk) [inline pkg/channel/consolidated/status/status.go:201]: would checking if the set is `Equal()` work?

File: pkg/channel/consolidated/status/status.go
Comment (lberk) [inline pkg/channel/consolidated/status/status.go:207]: why're we doing the unlocking, cancelprobing, re-lock when it seems like CancelProbing does the same thing as lines 214-216?

I'd prefer an approach where we're not unlocking, calling another function and then relocking like this tbh

File: pkg/channel/consolidated/status/status.go
Comment (devguyio) [reply pkg/channel/consolidated/status/status.go:192]: using single letter variable names for such extremely short lived variables is pretty idiomatic in go. I understand it's not readable if the variables would've had a bigger scope, but this is literally a single line scope where you'd know immediately what they from the assignment 3 lines above.

File: pkg/channel/consolidated/status/status.go
Comment (lberk) [inline pkg/channel/consolidated/status/status.go:517]: is adding a single use, single char variable like this for just the next line really more readable than just using `s.ready` ?

File: pkg/channel/consolidated/status/status.go
Comment (devguyio) [reply pkg/channel/consolidated/status/status.go:194]: because of the lock, accessing the map needs to happen under lock and this construct of lock, defer ulock, inside an anonymous function is both idiomatic and less bug prone (since unlocking is deferred)

**Comment by lberk** [comment]
there's quite a few new code paths that unit testing isn't covering.  were there any planned additions to unit tests?

File: pkg/channel/consolidated/status/status.go
Comment (devguyio) [reply pkg/channel/consolidated/status/status.go:355]: the one on line 365 was overlooked.

File: pkg/channel/consolidated/status/status.go
Comment (devguyio) [reply pkg/channel/consolidated/status/status.go:517]: nope, that function changed a lot and that one was overlooked

**Comment by devguyio** [comment]
@lberk yes and I created this issue to increase the unit tests coverage of the prober #491

File: pkg/channel/consolidated/status/status.go
Comment (devguyio) [reply pkg/channel/consolidated/status/status.go:194]: the whole idea is to rely on go's defer reliability which the go runtime will ensure to take place even if the call panics, something that the normal call to 'unlock' doesn't guarantee. Check the example by the end here https://blog.golang.org/defer-panic-and-recover

File: pkg/channel/consolidated/status/status.go
Comment (devguyio) [reply pkg/channel/consolidated/status/status.go:174]: `Len()` return type is `int`

File: pkg/channel/consolidated/status/status.go
Comment (devguyio) [reply pkg/channel/consolidated/status/status.go:355]: I changed the one on 365.

**Comment by devguyio** [comment]
@lberk if you've done reviewing, do you mind giving me a summary and what does this PR needs to be merged from your POV?

**Review by slinkydeveloper — COMMENTED** [review]
The idea sounds correct to me, but I'll let somebody else approve it :)

File: pkg/channel/consolidated/status/status.go
Comment (matzew) [reply pkg/channel/consolidated/status/status.go:194]: perhaps adding a comment, on why?

the debate here shows, there is reasons, but also questions.  Comments usually help to get larger context

**Comment by lberk** [comment]
> @lberk yes and I created this issue to increase the unit tests coverage of the prober #491

So basically, no.  There will be no futher unit test additions with this PR.

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (lberk) [inline pkg/channel/consolidated/reconciler/controller/controller.go:160]: If these two functions are essentially copy and pasted (save for a debug line).  Abstract the identical code into a different function and pass it with the debug string.

Comment (matzew) [reply pkg/channel/consolidated/reconciler/controller/controller.go:160]: I agree here, that `DeleteFunc` and `Addfunc` are very very identical - let's not start copy/paste here for no reason

**Comment by matzew** [comment]
> So basically, no. There will be no futher unit test additions with this PR.

@lberk I see where you coming from, and I think we should always try have coverage of code via unit tests. 

I lack context of why the unit tests for the `status.go` are part of a different issue/PR and can not be part of this PR...

But I am fine in adding that afterwards, FWIW.

File: pkg/channel/consolidated/status/status.go
Comment (lberk) [inline pkg/channel/consolidated/status/status.go:327]: could you elaborate to me why it's a sound design decision to have:
`CancelProbing` (which is part of the public Manager interface)
`CancelProbing`'s single job is to aquire the mutex lock, and call `forgetState(sub)`
`forgetState(sub)` has a comment a comment that indicates that we *must* have the lock before calling it
we then have other ways of calling `forgetState` elsewhere in the file.

Why aren't we using `CancelProbing` as the proper entrypoint my consistently?

**Comment by devguyio** [comment]
> So basically, no. There will be no futher unit test additions with this PR.

@lberk I misunderstood what you meant earlier by "were there any planned additions to unit tests?" I thought you meant "do you plan on adding unit tests later after this PR". And indeed I did, so I went ahead and created the issue to confirm that there's an issue and planned work for adding unit tests.

> I lack context of why the unit tests for the status.go are part of a different issue/PR and can not be part of this PR...

@matzew Reasons are: 

1. Time pressure.
2. We have some e2e tests in place that're giving us some level confidence.

**Comment by lberk** [comment]
> @matzew Reasons are:
> 
>     1. Time pressure.
>     2. We have some e2e tests in place that're giving us some level confidence.

Neither of these are valid criterion.

**Comment by devguyio** [comment]
@lberk I put it on hold to add unit tests and increase the coverage. Please let me know what else is needed for the PR to be approved? I'm addressing some comments too and responded to the others.

File: pkg/channel/consolidated/status/status.go
Comment (devguyio) [reply pkg/channel/consolidated/status/status.go:194]: sure, will add a comment

File: pkg/channel/consolidated/status/status.go
Comment (matzew) [reply pkg/channel/consolidated/status/status.go:327]: OK, so the issue is:

for outside invocation he have the `CancelProbe`, that basically does the lock and calls the private one.

Internally we call the `forget` since we have already taken care of the lock.

Perhaps the different names of `CancelProbe` versus `froget...` is confusing?

I think I understand the way of decorating the invocation of the `forget` with a lock, inside of the `CancelProbe`.

So...  Let's perhaps change the name?  `internalCancelProbe`, and be more explicit that it's only used w/in a mutex, and outside users should be instead calling `CancelProbe` ? 

And perhaps add a TODO... to have that sorted later

**Comment by lberk** [comment]
https://github.com/knative-sandbox/eventing-kafka/blob/9128624c5e5a7d0f4f684585bb25e98b8a909bb8/pkg/channel/consolidated/status/status.go#L110
is that ingress line still valid here?

**Comment by devguyio** [comment]
> is that ingress line still valid here?
Nope, it was overlooked

**Comment by devguyio** [comment]
@lberk @matzew I increased the test coverage , please check.
/unhold

**Comment by devguyio** [comment]
Checking the failed tests

File: pkg/channel/consolidated/status/status.go
Comment (lberk) [inline pkg/channel/consolidated/status/status.go]: just so I understand, why was this condition no longer needed?

Comment (devguyio) [reply pkg/channel/consolidated/status/status.go]: To force refreshing the Pods regardless of whether the IPs changed or not. We can't for sure know if pods with the same IPs are the same ones or not.

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (lberk) [inline pkg/channel/consolidated/reconciler/controller/controller.go:169]: :) small typo

File: pkg/channel/consolidated/reconciler/controller/controller.go
Comment (lberk) [inline pkg/channel/consolidated/reconciler/controller/controller.go:30]: kafkavmessaging1beta1 or kafkamessagingv1beta1

Comment (devguyio) [reply pkg/channel/consolidated/reconciler/controller/controller.go:30]: woops, typo. Good catch

File: pkg/channel/consolidated/status/status_test.go
Comment (lberk) [inline pkg/channel/consolidated/status/status_test.go:182]: hm, just wondering out loud here, should we be `defer close(probeRequestsDX` elsewhere as well?

**Comment by lberk** [comment]
/lgtm

Thanks for adding the tests @devguyio

File: pkg/channel/consolidated/status/status_test.go
Comment (devguyio) [reply pkg/channel/consolidated/status/status_test.go:182]: It wouldn't hurt, but it's not needed IMHO. The reason those channels are closed is to shutdown the prober and free the polling goroutines in the prober and not leak those routines. While closing the probeRequests won't be necessary since we're closing the whole server anyway and that'll clean up their goroutines IIUC. And as you know, you don't need to close all the channels since they'll be garbage collected if unreachable.

If you feel I should do that, sure I can. Let me know.

**Comment by devguyio** [comment]
@lberk sure, np. Are you gonna wait for someone else to approve the PR?

**Review by lberk — APPROVED** [review]

**Review by lionelvillard — APPROVED** [review]

**Review by matzew — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
